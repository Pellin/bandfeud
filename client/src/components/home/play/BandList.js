import React, { Component } from 'react';

import BandItem from './BandItem';

class BandList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      left: '0px',
      children: 0
    };
  }
  componentDidMount = () => {
    window.addEventListener('resize', this.setCenter);
    this.setState(() => ({ children: this.props.bands.length }));
    this.setCenter();
  };
  componentDidUpdate = () => {
    if (this.props.bands.length !== this.state.children) {
      this.setState(() => ({
        children: this.props.bands.length
      }));
      this.setState(() => ({
        left: parseInt(this.state.left.slice(0, -2)) - 210 + 'px'
      }));
    }
  };
  componentWillUnmount = () => {
    window.removeEventListener('resize', this.setCenter);
  };
  setCenter = () => {
    this.setState(() => ({
      left: window.innerWidth / 2 - this.props.bands.length * 210 + 105 + 'px'
    }));
  };
  render() {
    return (
      <div
        style={{
          alignItems: 'center',
          display: 'flex',
          flexDirection: 'row',
          justifyContent: 'center',
          top: '5rem',
          marginBottom: '1rem',
          left: this.state.left,
          position: 'absolute',
          transition: 'left, 1s'
        }}
      >
        {this.props.bands.map(band => (
          <BandItem key={band.name} name={band.name} url={band.url} points={band.points}/>
        ))}
      </div>
    );
  }
}

export default BandList;
