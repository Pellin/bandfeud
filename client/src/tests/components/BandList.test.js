import React from 'react';
import { shallow } from 'enzyme';

import BandList from '../../components/home/play/BandList';

let oneBand, threeBands, sixBands;

beforeEach(() => {
  oneBand = [{ name: 'one', url: 'wwww.1img.com' }];
  threeBands = [
    { name: 'one', url: 'wwww.1img.com' },
    { name: 'two', url: 'wwww.2img.com' },
    { name: 'three', url: 'wwww.3img.com' }
  ];
  sixBands = [
    { name: 'one', url: 'wwww.1img.com' },
    { name: 'two', url: 'wwww.2img.com' },
    { name: 'three', url: 'wwww.3img.com' },
    { name: 'four', url: 'wwww.4img.com' },
    { name: 'five', url: 'wwww.5img.com' },
    { name: 'six', url: 'wwww.6img.com' }
  ];
});

it('should render component correctly with 0 bands', () => {
  const wrapper = shallow(<BandList bands={[]} />);

  expect(wrapper).toMatchSnapshot();
});

it('should render correct number of BandItem components when new bands are added', () => {
  let wrapper = shallow(<BandList bands={oneBand} />);
  expect(wrapper.find('BandItem').length).toBe(1);

  wrapper = shallow(<BandList bands={threeBands} />);
  expect(wrapper.find('BandItem').length).toBe(3);

  wrapper = shallow(<BandList bands={sixBands} />);
  expect(wrapper.find('BandItem').length).toBe(6);
});

it("should set 'left' property correctly", () => {
  // 0 bands
  let expectedLeft = 617;
  let wrapper = shallow(<BandList bands={[]} />);
  expect(wrapper.state()).toEqual({ left: expectedLeft + 'px', children: 0 });
  expect(wrapper).toMatchSnapshot();
  // 1 band
  expectedLeft = 617 - 210;
  wrapper = shallow(<BandList bands={oneBand} />);
  expect(wrapper.state()).toEqual({ left: expectedLeft + 'px', children: 1 });
  expect(wrapper).toMatchSnapshot();
  // 3 bands
  expectedLeft = 617 - 630;
  wrapper = shallow(<BandList bands={threeBands} />);
  expect(wrapper.state()).toEqual({ left: expectedLeft + 'px', children: 3 });
  expect(wrapper).toMatchSnapshot();
  // 6 bands
  expectedLeft = 617 - 1260;
  wrapper = shallow(<BandList bands={sixBands} />);
  expect(wrapper.state()).toEqual({ left: expectedLeft + 'px', children: 6 });
  expect(wrapper).toMatchSnapshot();
});
