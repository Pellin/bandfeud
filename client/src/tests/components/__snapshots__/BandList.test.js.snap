// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component correctly with 0 bands 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "center",
      "left": "617px",
      "marginBottom": "1rem",
      "position": "absolute",
      "top": "5rem",
      "transition": "left, 1s",
    }
  }
/>
`;

exports[`should set 'left' property correctly 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "center",
      "left": "617px",
      "marginBottom": "1rem",
      "position": "absolute",
      "top": "5rem",
      "transition": "left, 1s",
    }
  }
/>
`;

exports[`should set 'left' property correctly 2`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "center",
      "left": "407px",
      "marginBottom": "1rem",
      "position": "absolute",
      "top": "5rem",
      "transition": "left, 1s",
    }
  }
>
  <BandItem
    key="one"
    name="one"
    url="wwww.1img.com"
  />
</div>
`;

exports[`should set 'left' property correctly 3`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "center",
      "left": "-13px",
      "marginBottom": "1rem",
      "position": "absolute",
      "top": "5rem",
      "transition": "left, 1s",
    }
  }
>
  <BandItem
    key="one"
    name="one"
    url="wwww.1img.com"
  />
  <BandItem
    key="two"
    name="two"
    url="wwww.2img.com"
  />
  <BandItem
    key="three"
    name="three"
    url="wwww.3img.com"
  />
</div>
`;

exports[`should set 'left' property correctly 4`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "center",
      "left": "-643px",
      "marginBottom": "1rem",
      "position": "absolute",
      "top": "5rem",
      "transition": "left, 1s",
    }
  }
>
  <BandItem
    key="one"
    name="one"
    url="wwww.1img.com"
  />
  <BandItem
    key="two"
    name="two"
    url="wwww.2img.com"
  />
  <BandItem
    key="three"
    name="three"
    url="wwww.3img.com"
  />
  <BandItem
    key="four"
    name="four"
    url="wwww.4img.com"
  />
  <BandItem
    key="five"
    name="five"
    url="wwww.5img.com"
  />
  <BandItem
    key="six"
    name="six"
    url="wwww.6img.com"
  />
</div>
`;
